Drag&Drop for FileInput

jhadvig opened the pull request:
Updated the `FileInput` component with having Drag&Drop functionality using the `react-dropzone` pkg.
Screen:
![1](https://user-images.githubusercontent.com/1668218/41986684-222b36ea-7a37-11e8-99c7-905dbcc04640.png)

Originally I wanted to create a HOC component that would wrap passed component with D&D functionality, but that turned out to be a bit more complicated since the data (fileName, fileData) needed to be passed back and forth. But I've figured out that the FileInput should have this functionality by default. So until we will need other components that would need D&D functionality, would leave this as is.

![create source secret tectonic](https://user-images.githubusercontent.com/1668218/41987058-67493780-7a38-11e8-8931-4767509bae65.gif)

@spadgett @openshift/team-ux-review  PTAL

alecmerdler commented on frontend/package.json:
```diff
@@ -73,6 +73,7 @@
     "react-dnd": "^2.6.0",
     "react-dnd-html5-backend": "^2.6.0",
     "react-dom": "16.x",
+    "react-dropzone": "4.x",
```
Are you concerned about using a module that is [not actively maintained](https://github.com/react-dropzone/react-dropzone/issues/479)?

jhadvig replied:
The issue is a bit misleading, since the PRs are being reviewed and merged. When you check the conversation, its about drafting additional reviewers, since the main contributer is there alone.

spadgett replied:
It looks like the existing react-dnd library we have can handle file drops. Can you investigate?

Would be good to avoid having two libraries that can do the same thing.

spadgett replied:
I'm looking at

http://react-dnd.github.io/react-dnd/docs-drop-target.html#handling-files-and-urls

> When using the HTML5 backend, you can handle the file drops just by registering a drop target for HTML5Backend.NativeTypes.FILE or HTML5Backend.NativeTypes.URL built-in types.

alecmerdler commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,16 +1,26 @@
 import * as React from 'react';
+import Dropzone from 'react-dropzone';
```
Can you make sure usages of this module use the `AsyncComponent` pattern to ensure our bundle size isn't increased just for this one dependency?

spadgett commented:
/retest

spadgett commented on frontend/public/components/utils/_file-input.scss:
```diff
@@ -15,5 +15,30 @@
+  bottom: 0;
+  left: 0;
+  background: rgba(0,0,0,0.5);
+  z-index: 1001;
```
nit: alpha (here and below)

spadgett commented on frontend/public/components/utils/_file-input.scss:
```diff
@@ -15,5 +15,30 @@
+}
+
+.help-block {
+  margin-bottom: 10px;
```
Why change the default help-block margin?

spadgett commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -21,19 +31,57 @@ export class FileInput extends React.Component<FileInputProps, FileInputState> {
+  activeDragZone() {
+    this.setState({ dropzoneActive: true });
+  }
+  deactiveDragZone() {
```
I'd stick with onDragEnter and onDragLeave for names

spadgett commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -21,19 +31,57 @@ export class FileInput extends React.Component<FileInputProps, FileInputState> {
+              id={id}
+              onChange={this.onDataChange}
+              value={this.state.inputFileData}
+              aria-describedby={`${id}-help`}
```
Looks like this should be

```html
aria-describedby="ssh-privatekey-help"
```

jhadvig replied:
No the `help-block` paragraph below should be set to `id={`${id}-help`}`

spadgett commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -21,19 +31,57 @@ export class FileInput extends React.Component<FileInputProps, FileInputState> {
   }
+  onFileDrop(event) {
+    this.setState({ dropzoneActive: false });
+    this.readFile(event[0]);
```
Can you double check there's nothing else we need to do here?

https://github.com/react-dropzone/react-dropzone#word-of-caution-when-working-with-previews

> Important: react-dropzone doesn't manage dropped files. You need to destroy the object URL yourself whenever you don't need the preview by calling window.URL.revokeObjectURL(file.preview); to avoid memory leaks.

I think we might need `disablePreview` if we're not using previews.

robszumski commented:
The grey box seems really intense to me. Brainstorming possible solutions:

Maybe tone down the bgcolor to be this lighter grey? Bringing down the opacity should get you really close to this grey.
And maybe a thicker border in `pf-blue` similar to below to indicate the border that you can drop things into? It should scale to any size pretty well.

Thoughts @openshift/team-ux-review?

![image](https://user-images.githubusercontent.com/755540/42117160-03065c20-7bb0-11e8-91df-a323bbac4c2d.png)


Basically combine both of these into one:

![image](https://user-images.githubusercontent.com/755540/42117038-5da41786-7baf-11e8-92d8-4dd6c9fae338.png)
![image](https://user-images.githubusercontent.com/755540/42117112-aee1f992-7baf-11e8-9077-8855be42d853.png)

jhadvig commented:
@robszumski here is the current visual of the DnD area:


![create source secret tectonic](https://user-images.githubusercontent.com/1668218/42168185-93cefb06-7e10-11e8-83ab-f57cff97f8a6.gif)


I've used dashed border and also the text color is changing upon hover. PTAL

cshinn commented:
@jhadvig what causes the "drag file here" box to appear? I can't really tell from the GIFs. Is it visible until something is entered in the text box, or when the user hovers over it?

jhadvig commented:
@cshinn sorry for got to add info since the mouse is not visible, while dragging the file. 
So the "Drag File Here" box appears when the dragged file hovers over the console. When the dragged file hovers over the drop-able area the border and font color will turn blue.

cshinn commented:
@jhadvig in that case, I think we need something to tell people that they can drag and drop files into the editor. Maybe change the blurb above the text box to something like:

"Drag and drop your private SSH key here or browse to upload it"

beanh66 commented:
@jhadvig We are currently working on a PatternFly contribution for all file upload cases, including this use case with the code editor. @bmignano is just about done with the design, so wondering if we can mimic the same styles. This is similar to what @robszumski proposed with the blue outline. See below: 
![file](https://user-images.githubusercontent.com/18057044/42325474-60b73596-8034-11e8-98cf-c6fe2e6f7954.png)
![drag-file](https://user-images.githubusercontent.com/18057044/42325480-65260328-8034-11e8-817b-352ce81f2fad.png)

jhadvig commented:
@beanh66 in the first wireframe, is the "Drop a file, paste in text or browse to upload" text in the editor by default? or it appears during when user start DnD over the `window` ?

cshinn commented:
@jhadvig I believe it is intended to appear there when the editor is empty

jhadvig commented:
@spadgett I've updated the PR, based on your conversation.
wrapped the `App` component with DragDropContext so we should be safe now, when reusing any component that should support DnD. Added as separate commit.

Also updated styling based on the conversation with @cshinn && @beanh66 

![1](https://user-images.githubusercontent.com/1668218/42397735-488e6744-8167-11e8-9625-1bd56a426405.gif)

PTAL

spadgett commented on frontend/public/components/app.jsx:
```diff
@@ -136,82 +139,85 @@ class App extends React.PureComponent {
-        </Switch>
-      </div>
-    </React.Fragment>;
+    return <DragDropContextProvider backend={HTML5Backend}>
```
This will break async loading of the component, won't it?

jhadvig replied:
No, its not breaking the async loading. Tested it via the inspector on the name-value-editor and its loaded asynchronously, so this shouldn't effect it.

spadgett replied:
But is react-dnd loaded async? That's the dependency we're worried about.

Since we have an impoty in app.jsx now, I don't think it can be.

jhadvig replied:
I thought that we want to import the react-dnd in app.jsx, so we ensure the HTML5Backend is loaded only once, for the whole console.
But I can rework it ina  way it will be loaded async only in the components that will use DnD.

spadgett replied:
> I thought that we want to import the react-dnd in app.jsx, so we ensure the HTML5Backend is loaded only once, for the whole console.

Ideally we'd load react-dnd async since it's not small and only used in a couple places. If so, we should initialize the HTML5Backend once on first load.

The reason we made the name value editor async was to lazy load react-dnd.

jhadvig commented:
@spadgett I've updated the PR. Loading the `DragDropContext` with HTML5Backend async only on subforms that need the DnD functionality. Tested the change on multiple dropable zones.
PTAL

spadgett commented on frontend/public/lib/withDragDropContext.tsx:
```diff
@@ -0,0 +1,6 @@
+
+// // Need to create this decorator, because of the issue with react-dnd module:
+// // https://github.com/react-dnd/react-dnd/issues/186#issuecomment-282789420
+// export default DragDropContext(HTML5Backend);
```
Realize this is probably still WIP, but we'll want to clean up some of this commented out code before merging.

spadgett commented on frontend/public/components/secrets/subforms.tsx:
```diff
@@ -0,0 +1,109 @@
+import HTML5Backend from 'react-dnd-html5-backend';
+import { DragDropContextProvider, DragDropContext } from 'react-dnd';
+
+export const SSHAuthSubform = DragDropContext(HTML5Backend)(class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubformState> {
```
Isn't this going to initialize the context multiple times?

spadgett commented on frontend/public/components/secrets/subforms.tsx:
```diff
@@ -0,0 +1,109 @@
+    }, () => this.props.onChange(this.state));
+  }
+  render() {
+    return <DragDropContextProvider backend={HTML5Backend}>
```
I didn't think you were supposed to use both `DragDropContext` and `DragDropContextProvider`

jhadvig replied:
we need both, at least based on the docs and examples

spadgett replied:
Can you point me to the doc or example? I'm looking at

> **As an alternative to the DragDropContext**, you can use the DragDropContextProvider element to set up React DnD for your application. Similar to the DragDropContext, this may be injected with a backend via the backend prop, but it also can be injected with a window object.

http://react-dnd.github.io/react-dnd/docs-drag-drop-context-provider.html

I could be wrong :)

jhadvig replied:
> Can you point me to the doc or example?
This is the [example](https://github.com/react-dnd/react-dnd/blob/master/packages/documentation/examples/06%20Other/Native%20Files/Container.tsx) I was digging from.

Looking on your docs once more :rocket:

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -293,24 +248,68 @@ class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubform
+  } else {
+    secretForm = withSecretForm(SourceSecretForm);
+  }
+  return withDragDropContext(secretForm);
```
@spadgett wrapping the returned component from the `withSecretForm` HOC into the `withDragDropContext` HOC... not sure we want to load the HTML5Context for all the secret forms though. Or lets load them only those that will actually used DnD ?

jhadvig commented:
@spadgett PR updated, based on our conversation. PTAL

spadgett commented:
Looks like you have an empty file: frontend/public/components/utils/drop-file.tsx

spadgett commented:
@jhadvig I think it's the DroppableFileInput component you want to load async, not the SSH form. I would

* Pull the SSH subform component back into create-secret. You shouldn't need the separate file.
* Use AsyncComponent to load the DroppableFileInput component for the source secret form

spadgett commented:
/hold
for feature freeze

jhadvig commented:
@spadgett I've updated the PR based on your comment. Makes more sense now :+1: 
tested thew changes and works with multiple dropable areas.
PTAL

spadgett commented:
jenkins rebuild

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -5,9 +5,11 @@ import { Helmet } from 'react-helmet';
 import { formatNamespacedRouteForResource } from '../../ui/ui-actions';
 import { WebHookSecretKey } from '../secret';
+import { AsyncComponent } from '../utils/async';
+import withDragDropContext from '../utils/withDragDropContext';
```
This is going to pull in react-dnd to the main bundle. I think this needs to be in file-input.tsx if we can move it.

jhadvig replied:
if we move it to file-input.tsx then we wont be able to have more then one instances of the `DropableFileInput` component, since everytime the HTML5Backend will be loaded.

jhadvig replied:
but you are right, we need to load it async

spadgett replied:
It's a singleton, though. If that doesn't work, then something is wrong in how we defined `withDragDropContext`

spadgett commented on frontend/public/components/utils/withDragDropContext.tsx:
```diff
@@ -0,0 +1,6 @@
+import {DragDropContext} from 'react-dnd';
```
I missed this on previous review, but JS/TS filenames should be kebab cased. Let's call this something like `drag-drop-context.tsx`.

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -293,24 +251,62 @@ class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubform
+  }
+});
+
+class SourceSecretForm extends React.Component<SourceSecretFormProps, SourceSecretFormState> {
```
Do you mind moving this back where it was to avoid the large diff? (Assuming you didn't move it here for some other reason.)

spadgett commented on frontend/public/components/utils/_file-input.scss:
```diff
@@ -15,5 +15,35 @@
+      border-color: $color-pf-blue-300;
+      color: $color-pf-blue-300;
+    }
+    .drop-text {
```
Better to use BEM naming for these classes. (I think we're setting the class names ourselves?)

spadgett commented:
Before:
> vendors~main: 786 KB

After:
> vendors~main: 887 KB

I wouldn't expect to see this increase if we have the async component working correctly. @alecmerdler I'm assuming that's not expected?

alecmerdler commented:
That is true. Try using `yarn run analyze` and make check to make sure your code is in a separate bundle with it's own `node_modules`.

jhadvig commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,151 @@
-  onFileChange(event) {
-    const file = event.target.files[0];
-    this.setState({ inputFileName: file.name });
+  readFile(file) {
```
https://jira.coreos.com/browse/CONSOLE-635

jhadvig commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,151 @@
+              </div>
+              <p className="help-block" id={`${id}-help`}>{this.props.inputFieldHelpText}</p>
+              <textarea className="form-control co-create-secret-form__textarea"
+                onChange={this.onDataChange}
```
Yeah was thinking about this, but figured out that it could be handy to see the which file was uploaded, and even if user decides to change some parts of the uploaded file, he will still see which file he is based from

jhadvig commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,151 @@
+    const id = this.props.id;
+    const { connectDropTarget, isOver, canDrop } = this.props;
+    return (
+      connectDropTarget &&
```
yeah I dont think there would be a case when the `connectDropTarget` won't be set, since it returns `connect.dropTarget()`

jhadvig commented on frontend/public/components/utils/_file-input.scss:
```diff
@@ -15,5 +15,35 @@
     right: 0;
     text-align: right;
     top: 0;
+    z-index: 1000;
```
This one can go away, since the position is absolute

jhadvig commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,151 @@
+    this.setState({
+      inputFileData: fileData
+    }, () => this.props.onChange(fileData));
+    !_.isEmpty(fileName) && this.setState({inputFileName: fileName});
```
ok, I've put it into one call, this way the filename will cleared when user will modify the uploaded file.

jhadvig commented:
@spadgett comments addressed. I've fixed the issue with the module size. 
Turned out that utils/index.js contained file-input.tsx, which loaded it by default. Removed it and should be right now.

jhadvig commented:
/retest

spadgett commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,154 @@
+  }
+  render() {
+    const id = this.props.id;
+    const { connectDropTarget, isOver, canDrop } = this.props;
```
You can add `id` here instead of setting it above

jhadvig commented:
@spadgett comment addressed, added a separate commit for it... Will squash once we are ready to merge :)

spadgett commented:
I'd go ahead and squash. It might make rebases easier if needed, and we'll squash before merging anyway.

jhadvig commented:
@spadgett squashed

spadgett commented:
/lgtm

jhadvig commented:
@spadgett rebased

spadgett commented:
/hold cancel

spadgett commented:
/lgtm

jhadvig commented:
jenkins rebuild

jhadvig commented:
/retest

jhadvig commented:
/retest

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

jhadvig commented:
/retest

spadgett commented:
@jhadvig /retest re-runs the prow tests, but not the jenkins ones

jhadvig commented:
@spadgett sorry for that :+1:

jhadvig commented:
jenkins rebuild

spadgett commented:
> sorry for that

Yeah, it's definitely not obvious. We should look at moving off Jenkins so everything is run through prow

spadgett commented:
jenkins rebuild

spadgett commented:
GUI-Tests-ALM has been renamed, so that test result is outdated. Merging manually since all tests have passed.
